elements/pl-multiple-choice/pl-multiple-choice.py

```diff
@@ -30,34 +28,134 @@ def prepare(element_html, data):
             else:
                 incorrect_answers.append(answer_tuple)
             index += 1
+    return correct_answers, incorrect_answers
+
+
+def prepare(element_html, data):
+    element = lxml.html.fragment_fromstring(element_html)
+    required_attribs = ['answers-name']
+    optional_attribs = ['weight', 'number-answers', 'fixed-order', 'inline', 'none-of-the-above', 'all-of-the-above', 'hide-letter-keys']
+    pl.check_attribs(element, required_attribs, optional_attribs)
+    name = pl.get_string_attrib(element, 'answers-name')
+
+    correct_answers, incorrect_answers = categorize_options(element)
 
     len_correct = len(correct_answers)
     len_incorrect = len(incorrect_answers)
     len_total = len_correct + len_incorrect
 
-    if len_correct < 1:
-        raise Exception('pl-multiple-choice element must have at least one correct answer')
-
-    number_answers = pl.get_integer_attrib(element, 'number-answers', len_total)
+    enable_nota = pl.get_boolean_attrib(element, 'none-of-the-above', NONE_OF_THE_ABOVE_DEFAULT)
+    enable_aota = pl.get_boolean_attrib(element, 'all-of-the-above', ALL_OF_THE_ABOVE_DEFAULT)
+
+    nota_correct = False
+    aota_correct = False
+    if enable_nota or enable_aota:
+        prob_space = len_correct + enable_nota + enable_aota
+        rand_int = random.randint(1, prob_space)
+        # Either 'None of the above' or 'All of the above' is correct
+        # with probability 1/(number_correct + enable-nota + enable-aota).
+        # However, if len_correct is 0, nota_correct is guaranteed to be True.
+        # Thus, if no correct option is provided, 'None of the above' will always
+        # be correct, and 'All of the above' always incorrect
+        nota_correct = enable_nota and (rand_int == 1 or len_correct == 0)
+        # 'All of the above' will always be correct when no incorrect option is
+        # provided, while still never both True
+        aota_correct = enable_aota and (rand_int == 2 or len_incorrect == 0) and not nota_correct
+
+    if len_correct < 1 and not enable_nota:
+        # This means the code needs to handle the special case when len_correct == 0
+        raise Exception('pl-multiple-choice element must have at least 1 correct answer or set none-of-the-above')
+
+    if enable_aota and len_correct < 2:
+        # To prevent confusion on the client side
+        raise Exception('pl-multiple-choice element must have at least 2 correct answers when all-of-the-above is set')
+
+    # 1. Pick the choice(s) to display
+    number_answers = pl.get_integer_attrib(element, 'number-answers', -1)
+    # determine if user provides number-answers
+    set_num_answers = True
+    if number_answers == -1:
```

Comment:
```suggestion
    number_answers = pl.get_integer_attrib(element, 'number-answers', None)
    # determine if user provides number-answers
    set_num_answers = True
    if number_answers is None:
```

let's be more explicit here and use `None` instead of `-1`